File: python/detect_video.py
import cv2
import time
import argparse
import numpy as np

# Import the necessary classes from your existing environment
from common import (
    HailoPythonInferenceEngine, 
    DetectionPostProcessor, 
    scale_detections_to_original
)

def preprocess_frame(frame, target_size=(640, 640), normalize=False):
    """
    Real-time preprocessing for a raw OpenCV frame.
    Handles letterbox resizing, BGR to RGB conversion, and batch dimension padding.
    """
    orig_h, orig_w = frame.shape[:2]
    
    # Calculate scaling factor to fit within target size while maintaining aspect ratio
    scale = min(target_size[0] / orig_h, target_size[1] / orig_w)
    new_h, new_w = int(orig_h * scale), int(orig_w * scale)
    
    # Resize image
    resized = cv2.resize(frame, (new_w, new_h), interpolation=cv2.INTER_LINEAR)
    
    # Calculate padding to reach target size
    pad_h = (target_size[0] - new_h) // 2
    pad_w = (target_size[1] - new_w) // 2
    
    # Create canvas and place resized image inside
    canvas = np.full((target_size[0], target_size[1], 3), 114, dtype=np.uint8)
    canvas[pad_h:pad_h+new_h, pad_w:pad_w+new_w] = resized
    
    # CRITICAL: Swap BGR (OpenCV default) to RGB (YOLO training default)
    canvas = cv2.cvtColor(canvas, cv2.COLOR_BGR2RGB)
    
    if normalize:
        canvas = canvas.astype(np.float32) / 255.0
        
    # Add batch dimension: (1, H, W, 3)
    input_data = np.expand_dims(canvas, axis=0)
    
    return input_data, (orig_h, orig_w), scale, pad_w, pad_h


def detect_video_stream(args):
    """Run real-time detection on a camera stream using Hailo-8L"""
    
    # Dynamically configure classes based on user input
    class_dict = {i: name for i, name in enumerate(args.labels)}
    DetectionPostProcessor.set_classes(class_dict)
    num_classes = len(args.labels)
    
    print(f"[Initializing Hailo Engine with {args.hef} for {num_classes} classes...]")
    engine = HailoPythonInferenceEngine(args.hef, num_classes=num_classes)
    print("✓ Engine loaded successfully.")

    # Open the video stream (handles device index or virtual video paths/IP cams)
    if args.camera_id.isdigit():
        cap = cv2.VideoCapture(int(args.camera_id))
    else:
        cap = cv2.VideoCapture(args.camera_id)
    
    if not cap.isOpened():
        print(f"Error: Could not open camera source: {args.camera_id}.")
        return

    # Set camera capture resolution
    cap.set(cv2.CAP_PROP_FRAME_WIDTH, args.input_width)
    cap.set(cv2.CAP_PROP_FRAME_HEIGHT, args.input_height)

    print(f"[Starting video stream ({args.input_width}x{args.input_height}). Press 'q' to quit.]")
    
    # FPS tracking variables
    frame_count = 0
    start_time = time.time()
    target_size = (args.model_size, args.model_size)
        
    while True:
        ret, frame = cap.read()
        if not ret:
            print("Error: Failed to grab frame from camera.")
            break
            
        orig_h, orig_w = frame.shape[:2]
        
        # 1. Preprocess the raw frame
        input_data, orig_size, scale, pad_w, pad_h = preprocess_frame(
            frame, 
            target_size=target_size, 
            normalize=args.normalize
        )
        
        # 2. Run inference
        t_infer_start = time.perf_counter()
        results, stats = engine.infer(input_data, verbose=False, save_output=False, conf_threshold=args.conf_threshold)
        infer_time = (time.perf_counter() - t_infer_start) * 1000 # in ms
        
        # 3. Post-process and scale boxes back to the original video resolution
        results = scale_detections_to_original(results, orig_h, orig_w, scale, pad_w, pad_h)
        
        # 4. Draw bounding boxes onto the frame
        output_frame = DetectionPostProcessor.draw_bboxes(frame, results, thickness=args.box_thickness)
        
        # 5. Calculate and display FPS and Inference Time
        frame_count += 1
        elapsed_time = time.time() - start_time
        fps = frame_count / elapsed_time
        
        # Overlay metrics on the top-left corner
        cv2.putText(output_frame, f"FPS: {fps:.1f}", (10, 30), cv2.FONT_HERSHEY_SIMPLEX, 0.6, (0, 0, 0), 2)
        cv2.putText(output_frame, f"Infer: {infer_time:.1f}ms", (10, 60), cv2.FONT_HERSHEY_SIMPLEX, 0.6, (0, 0, 0), 2)
                    
        # 6. Display the live feed
        cv2.imshow("Hailo-8L Real-Time YOLO26", output_frame)
        
        # Break the loop if the user presses 'q'
        key = cv2.waitKey(1) & 0xFF
        if key == ord('q'):
            break

    # Clean up hardware resources when the loop exits
    cap.release()
    cv2.destroyAllWindows()
    print("[Video stream closed.]")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Real-time USB Camera Detection with Hailo-8L + Python Head")
    
    # Hardware & Model Configuration
    parser.add_argument("--camera-id", type=str, default="0", 
                        help="Camera ID index (e.g., 0) or video file path / IP stream URL")
    parser.add_argument("--hef", type=str, required=True, 
                        help="Path to the compiled Hailo Executable Format (.hef) model")
    
    # Model Input Tuning
    parser.add_argument("--labels", type=str, nargs='+', default=['bottle'],
                        help="List of class names the model was trained on, separated by spaces (e.g., --labels car person).")
    parser.add_argument("--model-size", type=int, default=640, 
                        help="The input square dimensions the YOLO model expects (e.g., 640)")
    parser.add_argument("--normalize", action="store_true", 
                        help="Normalize input pixels to [0,1] float32 (omit if the HEF model takes raw uint8)")
    parser.add_argument("--conf-threshold", type=float, default=0.25, 
                        help="Confidence threshold for bounding box filtering")
    
    # Camera Stream Resolution
    parser.add_argument("--input-width", type=int, default=640, 
                        help="Set frame width parameter for camera hardware capture")
    parser.add_argument("--input-height", type=int, default=480, 
                        help="Set frame height parameter for camera hardware capture")
    
    # Visualization Tuning
    parser.add_argument("--box-thickness", type=int, default=2, 
                        help="Line thickness for bounding boxes drawn over the frames")
    
    args = parser.parse_args()
    detect_video_stream(args)

File: cpp/src/detect_video.cpp
/**
 * detect_video.cpp
 * * Real-time video stream detection with Hailo-8L + Post-processing in C++.
 * Features dynamic class labeling, configurable model sizes, and standard argparse.
 */

#include "hailo/hailort.hpp"
#include "hailo/vdevice.hpp"
#include "hailo/infer_model.hpp"
#include "hailo/vstream.hpp"

#include <opencv2/opencv.hpp>
#include <iostream>
#include <vector>
#include <string>
#include <chrono>
#include <map>
#include <cmath>
#include <algorithm>
#include <sstream>

#include "preprocess.hpp"
#include "postprocess.hpp"

using namespace hailort;

// ============================================================================
// Command Line Argument Parser Helper
// ============================================================================
class InputParser {
public:
    InputParser(int &argc, char **argv) {
        for (int i = 1; i < argc; ++i)
            this->tokens.push_back(std::string(argv[i]));
    }
    const std::string& getCmdOption(const std::string &option, const std::string &default_val = "") const {
        std::vector<std::string>::const_iterator itr;
        itr = std::find(this->tokens.begin(), this->tokens.end(), option);
        if (itr != this->tokens.end() && ++itr != this->tokens.end()) {
            return *itr;
        }
        static const std::string empty_string("");
        return default_val.empty() ? empty_string : default_val;
    }
    bool cmdOptionExists(const std::string &option) const {
        return std::find(this->tokens.begin(), this->tokens.end(), option) != this->tokens.end();
    }
private:
    std::vector<std::string> tokens;
};

// Helper to parse comma-separated labels
std::vector<std::string> parse_labels(const std::string& labels_str) {
    std::vector<std::string> labels;
    std::stringstream ss(labels_str);
    std::string item;
    while (std::getline(ss, item, ',')) {
        labels.push_back(item);
    }
    return labels;
}

// Helper to print errors
void print_error(hailo_status status, const std::string& msg) {
    std::cerr << "Error: " << msg << " (Status: " << status << ")" << std::endl;
}

// ============================================================================
// Main Execution Pipeline
// ============================================================================
int main(int argc, char** argv) {
    InputParser input(argc, argv);

    if (input.cmdOptionExists("--help") || input.cmdOptionExists("-h")) {
        std::cout << "Usage: ./detect_video [OPTIONS]\n"
                  << "Options:\n"
                  << "  --hef <path>             Path to compiled HEF model (Default: ../models/yolo26n.hef)\n"
                  << "  --camera-id <id/url>     USB camera index or IP stream URL (Default: 0)\n"
                  << "  --labels <l1,l2>         Comma-separated list of class names (Default: bottle)\n"
                  << "  --model-size <int>       Model input square size (Default: 640)\n"
                  << "  --input-width <int>      Camera capture width (Default: 640)\n"
                  << "  --input-height <int>     Camera capture height (Default: 480)\n"
                  << "  --conf-threshold <float> Confidence threshold (Default: 0.25)\n";
        return 0;
    }

    // 1. Parse Arguments
    std::string hef_path = input.getCmdOption("--hef", "../models/yolo26n.hef");
    std::string camera_id = input.getCmdOption("--camera-id", "0");
    std::string labels_str = input.getCmdOption("--labels", "bottle");
    int model_size = std::stoi(input.getCmdOption("--model-size", "640"));
    int input_width = std::stoi(input.getCmdOption("--input-width", "640"));
    int input_height = std::stoi(input.getCmdOption("--input-height", "480"));
    float conf_threshold = std::stof(input.getCmdOption("--conf-threshold", "0.25"));

    std::vector<std::string> class_names = parse_labels(labels_str);

    std::cout << "[Initializing Hailo Engine for " << class_names.size() << " classes...]" << std::endl;
    
    // 2. Setup Hailo Hardware Device
    auto vdevice_exp = VDevice::create();
    if (!vdevice_exp) { print_error(vdevice_exp.status(), "Failed to create VDevice"); return 1; }
    auto vdevice = std::move(vdevice_exp.value());

    auto hef_exp = Hef::create(hef_path);
    if (!hef_exp) { print_error(hef_exp.status(), "Failed to load HEF"); return 1; }
    auto hef = std::move(hef_exp.value());

    auto configure_params_exp = vdevice->create_configure_params(hef);
    auto configure_params = configure_params_exp.value();

    auto network_groups_exp = vdevice->configure(hef, configure_params);
    auto network_groups = std::move(network_groups_exp.value());
    auto network_group = network_groups[0];

    auto input_vstream_params = network_group->make_input_vstream_params(false, HAILO_FORMAT_TYPE_UINT8, HAILO_DEFAULT_VSTREAM_TIMEOUT_MS, HAILO_DEFAULT_VSTREAM_QUEUE_SIZE).value();
    auto output_vstream_params = network_group->make_output_vstream_params(false, HAILO_FORMAT_TYPE_FLOAT32, HAILO_DEFAULT_VSTREAM_TIMEOUT_MS, HAILO_DEFAULT_VSTREAM_QUEUE_SIZE).value();

    auto input_vstreams_exp = VStreamsBuilder::create_input_vstreams(*network_group, input_vstream_params);
    auto input_vstreams = std::move(input_vstreams_exp.value());

    auto output_vstreams_exp = VStreamsBuilder::create_output_vstreams(*network_group, output_vstream_params);
    auto output_vstreams = std::move(output_vstreams_exp.value());

    // Prepare output buffers map
    std::map<std::string, std::vector<float>> output_buffers;
    for (auto& ov : output_vstreams) {
        output_buffers[ov.name()] = std::vector<float>(ov.get_frame_size() / sizeof(float));
    }

    std::cout << "✓ Hailo Engine loaded successfully." << std::endl;

    // 3. Open Video Camera Stream
    cv::VideoCapture cap;
    if (camera_id.find_first_not_of("0123456789") == std::string::npos) {
        cap.open(std::stoi(camera_id)); // Open by device index
    } else {
        cap.open(camera_id); // Open by stream URL
    }

    if (!cap.isOpened()) {
        std::cerr << "Error: Could not open camera source: " << camera_id << std::endl;
        return -1;
    }

    cap.set(cv::CAP_PROP_FRAME_WIDTH, input_width);
    cap.set(cv::CAP_PROP_FRAME_HEIGHT, input_height);
    cap.set(cv::CAP_PROP_BUFFERSIZE, 1);

    std::cout << "[Starting video stream (" << input_width << "x" << input_height << "). Press 'q' to quit.]" << std::endl;

    cv::Mat frame;
    int frame_count = 0;
    auto start_time = std::chrono::steady_clock::now();
    
    while (true) {
        cap >> frame;
        if (frame.empty()) break;

        // A. Preprocess
        LetterboxInfo lb_info = letterbox_image(frame, model_size, model_size);
        cv::Mat resized_image = lb_info.image;
        cv::cvtColor(resized_image, resized_image, cv::COLOR_BGR2RGB);

        auto t_infer_start = std::chrono::steady_clock::now();

        // B. Run Inference
        input_vstreams[0].write(MemoryView(resized_image.data, resized_image.total() * resized_image.elemSize()));
        
        std::vector<const float*> cls_ptrs(3, nullptr);
        std::vector<const float*> reg_ptrs(3, nullptr);
        int num_classes = class_names.size();

        for (auto& ov : output_vstreams) {
            ov.read(MemoryView(output_buffers[ov.name()].data(), output_buffers[ov.name()].size() * sizeof(float)));
            
            // Map tensors safely using Hardware Shape Metadata instead of raw buffer sizes
            auto shape = ov.get_info().shape;
            if (shape.features == num_classes) { // Classification Heads
                if (shape.height == 80) cls_ptrs[0] = output_buffers[ov.name()].data();
                else if (shape.height == 40) cls_ptrs[1] = output_buffers[ov.name()].data();
                else if (shape.height == 20) cls_ptrs[2] = output_buffers[ov.name()].data();
            } else if (shape.features == 4) { // Regression Heads (BBox coords)
                if (shape.height == 80) reg_ptrs[0] = output_buffers[ov.name()].data();
                else if (shape.height == 40) reg_ptrs[1] = output_buffers[ov.name()].data();
                else if (shape.height == 20) reg_ptrs[2] = output_buffers[ov.name()].data();
            }
        }

        auto t_infer_end = std::chrono::steady_clock::now();
        double infer_time_ms = std::chrono::duration<double, std::milli>(t_infer_end - t_infer_start).count();

        // Safety verification
        if (!cls_ptrs[0] || !cls_ptrs[1] || !cls_ptrs[2] || !reg_ptrs[0] || !reg_ptrs[1] || !reg_ptrs[2]) {
            std::cerr << "Error: Could not map HEF output tensors based on shape metadata!" << std::endl;
            break;
        }

        // C. C++ Post-Processing
        std::vector<Detection> results = run_postprocess(
            IntList<8, 16, 32>{}, IntList<80, 40, 20>{},
            cls_ptrs, reg_ptrs, conf_threshold, num_classes
        );

        // D. Draw Detections
        for (const auto& det : results) {
            // Reverse the letterbox padding to align with original camera frame
            float x1_f = (det.x1 - lb_info.pad_w) / lb_info.scale;
            float y1_f = (det.y1 - lb_info.pad_h) / lb_info.scale;
            float x2_f = (det.x2 - lb_info.pad_w) / lb_info.scale;
            float y2_f = (det.y2 - lb_info.pad_h) / lb_info.scale;
            
            // Safety Clipping
            int x1 = std::max(0, std::min((int)x1_f, frame.cols));
            int y1 = std::max(0, std::min((int)y1_f, frame.rows));
            int x2 = std::max(0, std::min((int)x2_f, frame.cols));
            int y2 = std::max(0, std::min((int)y2_f, frame.rows));
            
            // Draw bounding box
            cv::Rect rect(cv::Point(x1, y1), cv::Point(x2, y2));
            cv::rectangle(frame, rect, cv::Scalar(255, 0, 0), 2);
            
            // Dynamically lookup label
            std::string class_name_str = (det.cls_id < class_names.size()) ? class_names[det.cls_id] : "Class_" + std::to_string(det.cls_id);
            char label[64];
            snprintf(label, sizeof(label), "%s (%.2f)", class_name_str.c_str(), det.conf);
            
            int font_face = cv::FONT_HERSHEY_TRIPLEX;
            double font_scale = 0.5;
            int thickness = 1;

            int baseline = 0;
            cv::Size text_size = cv::getTextSize(label, font_face, font_scale, thickness, &baseline);

            int text_y = rect.y - 5; 
            if (text_y - text_size.height < 0) {
                text_y = rect.y + text_size.height + 5; 
            }
            int text_x = std::max(0, rect.x);

            cv::Rect text_bg(text_x, text_y - text_size.height, text_size.width, text_size.height + baseline);
            cv::rectangle(frame, text_bg, cv::Scalar(255, 0, 0), cv::FILLED);
            cv::putText(frame, label, cv::Point(text_x, text_y), font_face, font_scale, cv::Scalar(0, 255, 0), thickness);
        }

        // E. Calculate and Draw FPS
        frame_count++;
        auto current_time = std::chrono::steady_clock::now();
        double elapsed_sec = std::chrono::duration<double>(current_time - start_time).count();
        double fps = frame_count / elapsed_sec;

        cv::putText(frame, "FPS: " + std::to_string(fps).substr(0, 4), cv::Point(10, 30), cv::FONT_HERSHEY_SIMPLEX, 0.6, cv::Scalar(0, 0, 0), 2);
        cv::putText(frame, "Infer: " + std::to_string(infer_time_ms).substr(0, 4) + "ms", cv::Point(10, 60), cv::FONT_HERSHEY_SIMPLEX, 0.6, cv::Scalar(0, 0, 0), 2);
        
        cv::imshow("Hailo C++ Real-Time Detector", frame);
        
        // Press 'q' to quit
        int key = cv::waitKey(1) & 0xFF;
        if (key == 'q') {
            break;
        }
    }
    
    cap.release();
    cv::destroyAllWindows();
    std::cout << "[Video stream closed.]" << std::endl;
    return 0;
}